### PR TITLE
feat(resources): add column filter inversion

### LINKED ISSUE #923: Allow search filter inversion
## Problem statement

I often want to check upon which pods have issues. For that I go to the "Resources" tab and filter for Pods that are **not** in the `Running` state. Since the list of filters is built dynamically on the available states, I have to go back to the filter and adjust it when ever pods enter new states.

## Proposed solution

Provide a flag to click to invert a filter. E.g., let me select `Running` and `Invert` to filter for all pods/resources that are not in the selected state(s).

## Alternatives considered

Manually adjust the filter every so often.

## Additional context

This is especially useful in namespaces with a lot of resource.

### PR DESCRIPTION
**Pull request by emanuelbesliu**
## Description

Adds an **Invert** toggle to Resources-tab column filters. Selecting values like `status=Running` and enabling Invert flips the filter to show every row that does **not** match — e.g. "show me all pods that are NOT Running" without re-editing the filter each time states change.

- New per-column invert state, persisted in the URL via a `filterInvert` param (round-trip serialize/parse helpers live in `resource-utils`).
- Invert is dropped automatically when a column's values are cleared, so a bare flag never lingers as URL noise or silently re-arms when values are reselected.
- The active-filter pill shows the negated form (`not …`) when a column is inverted.
- Invert-only (no selected values) is a no-op and doesn't count as an active filter.

## Type of change

- [x] New feature (non-breaking change that adds functionality)

## How has this been tested?

- [x] Tested locally with minikube/kind
- [x] Added/updated unit tests

Details:
- Added 4 round-trip unit tests for the invert serialization helpers in `column-filter-serialization.test.ts`.
- `make tsc` passes; `k8s-ui` vitest (1255 tests) and `web` vitest (161 tests) pass; production frontend build succeeds.
- Verified in the UI against a `kind` cluster: filtering `status=Running` (9 pods) then enabling Invert correctly shows only the single non-Running pod, and the URL serializes to `?filters=status:Running&filterInvert=status`.

## Checklist

- [x] My code follows the project's coding standards
- [x] I have performed a self-review of my code
- [x] I have added comments where necessary
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged

## Related issues

Fixes #923

---

> [!NOTE]
> **Low Risk**
> Changes are limited to client-side table filtering and URL serialization, with added guards on URL-derived object keys; no server or auth impact.
> 
> **Overview**
> Resources table column filters can now run in **Exclude** mode so rows matching the selected values are hidden instead of shown. Each column’s dropdown gets **Include** / **Exclude** toggles; active pills and header badges reflect negation (`not …`, `≠` count).
> 
> Exclude mode is stored in the existing `filters` query string via an `exclude` operator on each column (e.g. `status:exclude:Running`), with parse/serialize helpers extended in `resource-utils` and round-trip tests added. Exclude flags are cleared when a column has no values, when filters are reset, or when switching resource kinds; the node shortcut to filter pods by node always forces include for that column.
> 
> URL parsing now shares a pair parser and rejects prototype-polluting keys when building filter maps from query params.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1ff82bb81135c47c3c70e7359304ac0425a694ca. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by nadaverell** [comment]
Commented also on the issue - Thanks!
But regarding implementation - I'm not sure an "invert" toggle is the best UX here. WDYT about a filter-level toggle - change the existing component to optionally support Include/Exclude semantics.

Note - the active state should remain visibly negative after the popover closes, and the operator must persist in the URL while existing inclusive filter URLs continue to work.

For pod health specifically - perhaps Radar’s existing Problems classification should add an “Any problem” option rather than defining problems as “not Running”? That avoids including healthy Completed/transitional pods and catches unhealthy pods that still have a Running phase.

File: packages/k8s-ui/src/components/resources/ResourcesView.tsx [hunk 1]
```diff
@@ -2101,10 +2103,12 @@ function getInitialFiltersFromURL() {
   const params = new URLSearchParams(window.location.search)
   // Parse generic column filters
   const columnFilters = parseColumnFilters(params.get('filters'))
+  const columnFilterInverts = parseColumnFilterInverts(params.get('filterInvert'))
   const result = {
     search: params.get('search') || '',
     regex: params.get('regex') === 'true',
     columnFilters,
+    columnFilterInverts,
```
[the remark this answers is not in the document]
Comment (emanuelbesliu) [reply packages/k8s-ui/src/components/resources/ResourcesView.tsx]: Good catch — fixed in 40df8a4e. Parsed `filterInvert` keys are now reconciled against the active column filters in `getInitialFiltersFromURL` (via a new pure `reconcileColumnFilterInverts` helper), so a lone invert flag with no selected values is discarded instead of lying dormant. Added unit tests for the orphan/empty-values cases.

**Comment by emanuelbesliu** [comment]
Thanks for the great feedback @nadaverell — this is my first contribution to Radar, so I really appreciate it!

Totally agree: an Include/Exclude operator built into the shared filter component is nicer than a bolt-on "Invert" toggle. Happy to refactor that way, and I'll keep the three things you mentioned — the pill stays visibly negative after the popover closes, the operator persists in the URL, and existing inclusive filter URLs keep working.

One quick question: for the URL, do you prefer a separate additive param (like today's `filterInvert=status`) to keep old links untouched, or folding it into `filters` with an exclude prefix? I'll follow your lead.

On the pod-health idea — agree "not Running" is imprecise. That feels like a separate improvement to the Problems classification, so I'd suggest a follow-up issue to keep this PR focused. Happy to take that on.

**Comment by nadaverell** [comment]
> Thanks for the great feedback @nadaverell — this is my first contribution to Radar, so I really appreciate it!
> 
> Totally agree: an Include/Exclude operator built into the shared filter component is nicer than a bolt-on "Invert" toggle. Happy to refactor that way, and I'll keep the three things you mentioned — the pill stays visibly negative after the popover closes, the operator persists in the URL, and existing inclusive filter URLs keep working.
> 
> One quick question: for the URL, do you prefer a separate additive param (like today's `filterInvert=status`) to keep old links untouched, or folding it into `filters` with an exclude prefix? I'll follow your lead.
> 
> On the pod-health idea — agree "not Running" is imprecise. That feels like a separate improvement to the Problems classification, so I'd suggest a follow-up issue to keep this PR focused. Happy to take that on.

@emanuelbesliu you are most welcome to contribute to Radar - *we* appreciate it!

I think fold into each filter. Can do something like:

- Two parts: status:Running → default to include (backwards compatible)
- Three parts: status:exclude:Running → explicit mode

so both work:

?filters=status:Running
?filters=status:exclude:Running,Completed|namespace:include:default

And the UI can have maybe some toggle when opening the filter popup:
[ Include | Exclude ]

I don't think we need more complex conditions like "include Running but exclude Completed" so this shape seems like clean minimal change, easily discoverable right where you'd look.

**Comment by emanuelbesliu** [comment]
Done — pushed the refactor in 718132fc.

Went with your URL shape: the operator folds into each column entry in `filters`, so both forms work:

- `?filters=status:Running` (implicit include, unchanged)
- `?filters=status:exclude:Running,Completed|namespace:include:default`

Changes:
- Popover now has an `[ Include | Exclude ]` segmented toggle at the top; dropped the separate "Invert" checkbox and the old `filterInvert` param.
- Active-filter pill keeps the negated `not ...` form after the popover closes, and legacy inclusive URLs parse unchanged.
- A nice side effect: since the operator and its values share one entry, a stale operator with no values is now structurally impossible — that also resolves the orphan-flag edge case Bugbot flagged earlier.

Kept it to the two-state Include/Exclude per column (no mixed conditions), as you suggested. `make tsc` is clean and the k8s-ui suite passes (1260 tests, including new coverage for the include/exclude serialization).

Happy to open a separate issue for the Problems "Any problem" classification idea so it can be tracked on its own.

File: packages/k8s-ui/src/components/resources/ResourcesView.tsx [hunk 2]
```diff
@@ -4865,6 +4912,34 @@ export function ResourcesView({
...
+                                    'flex-1 px-2 py-1 text-xs rounded transition-colors',
+                                    columnFilterExcludes[col.key]
+                                      ? 'selection-strong selection-text'
+                                      : 'text-theme-text-secondary hover:bg-theme-elevated hover:text-theme-text-primary'
+                                  )}
+                                  title="Show rows that do NOT match the selected values"
+                                >
+                                  Exclude
+                                </button>
+                              </div>
```
[the remark this answers is not in the document]
Comment (emanuelbesliu) [reply packages/k8s-ui/src/components/resources/ResourcesView.tsx:4955]: Good catch — fixed in fb47212a. The node "filter by node" shortcut now reads the existing exclude map and passes it through `serializeColumnFilters`, so Exclude filters on other columns are preserved when jumping to a node's pods. Added coverage in the serialization tests.

File: packages/k8s-ui/src/components/resources/ResourcesView.tsx [hunk 3]
```diff
@@ -2248,6 +2275,14 @@ export function ResourcesView({
...
         if (updated.length === 0) {
           delete next[key]
+          // Inversion is meaningless with no selected values — drop it so a
+          // stale invert flag doesn't linger in the URL or re-arm on reselect.
+          setColumnFilterInverts(inv => {
+            if (!inv[key]) return inv
+            const nextInv = { ...inv }
+            delete nextInv[key]
+            return nextInv
+          })
```
[the remark this answers is not in the document]
Comment (emanuelbesliu) [reply packages/k8s-ui/src/components/resources/ResourcesView.tsx:2282]: Also handled in fb47212a. The refactor removed the invert map entirely, but the same stale-state risk applied to the new exclude map, so `removeCustomColumn` now clears `columnFilterExcludes[key]` alongside the values when a custom column is removed.

**Comment by nadaverell** [comment]
Much better, thanks. A few small things before merging:

- I think bugbot is right - adding a node filter needs to preserve the modes of existing filters (clicking a node when viewing pods list).
- Removing a custom column should also clear its entry from columnFilterExcludes
- optional but I think would be a nice touch - with the popover closed, we could show something different to distinguish Exclude from Include, like "≠ 1" instead of "1" maybe would work. Up to you.

**Review by nadaverell — APPROVED** [review]

**Comment by nadaverell** [comment]
LGTM - thanks again for taking this!
